### PR TITLE
More exception handling

### PR DESCRIPTION
**Pull request by pavanpodila**

### CODE REVIEW

**Comment by katis** [comment]
Hell of a job 🥇  Is this ready to merge?

**Comment by pavanpodila** [comment]
Not yet but you could leave one 👍. I've to fix coveralls as the coverage in the source file is missing due to the path change in mono repo. Plus a few other changes on exception handling

**Comment by pavanpodila** [comment]
@katis ready for review!

File: mobx/lib/src/api/action.dart [hunk 1]
```diff
@@ -34,21 +34,21 @@ import 'package:mobx/src/core.dart';
 /// can also be nested inside, in which case the change notification will propagate when
 /// the top-level action completes.
 Action action(Function fn, {String name, ReactiveContext context}) =>
-    Action(context ?? currentContext, fn, name: name);
+    Action(context ?? mainContext, fn, name: name);
```
Comment (pavanpodila) [inline mobx/lib/src/api/action.dart]: renamed to `mainContext` to suggest the primary context of the app.

File: mobx/lib/src/api/context.dart [hunk 2]
```diff
@@ -1,6 +1,18 @@
...
+final ReactiveContext mainContext = createContext(config: ReactiveConfig.main);
+
+/// Create a new context for running actions and reactions.
+///
+/// You can use this to run a reactivity system in parallel to the main app-level
+/// system. All actions, reactions will be run within this context. Make sure to pass
+/// this context in calls to autorun, reaction, when, action, observable, etc.
+///
+/// Most of the time you should be fine with the [mainContext]
+ReactiveContext createContext({ReactiveConfig config}) =>
```
Comment (pavanpodila) [inline mobx/lib/src/api/context.dart]: new API to create a context with some configuration

File: mobx/lib/src/core.dart [hunk 3]
```diff
@@ -27,4 +27,7 @@ class MobXCaughtException implements Exception {
   MobXCaughtException(exception) : _exception = exception;
 
   Object _exception;
+  Object get exception => _exception;
 }
+
+typedef Dispose = void Function();
```
Comment (pavanpodila) [inline mobx/lib/src/core.dart]: moved this from `utils.dart`

File: mobx/lib/src/core/computed.dart [hunk 4]
```diff
@@ -57,11 +61,15 @@ class ComputedValue<T> extends Atom implements Derivation {
     if (track) {
       value = _context.trackDerivation(this, _fn);
     } else {
-      try {
+      if (_context.config.disableErrorBoundaries == true) {
```
Comment (pavanpodila) [inline mobx/lib/src/core/computed.dart]: Ignore error boundary. This is part of the `ReactiveConfig`

File: mobx/lib/src/core/reaction_helper.dart [hunk 5]
```diff
@@ -146,24 +146,49 @@ ReactionDisposer createReaction<T>(ReactiveContext context,
 
 ReactionDisposer createWhenReaction(ReactiveContext context,
     bool Function(Reaction) predicate, void Function() effect,
-    {String name, void Function(Object, Reaction) onError}) {
+    {String name, int timeout, void Function(Object, Reaction) onError}) {
```
Comment (pavanpodila) [inline mobx/lib/src/core/reaction_helper.dart]: with the `timeout` option, when will throw error if the predicate does not execute within the given time

**Pull request merged**